tests/remotes.py

```diff
@@ -0,0 +1,325 @@
+    return url
+
+
+def get_hdfs_url():
+    return "hdfs://{}@127.0.0.1{}".format(
+        getpass.getuser(), get_local_storagepath()
+    )
+
+
+def get_aws_storagepath():
+    return TEST_AWS_REPO_BUCKET + "/" + str(uuid.uuid4())
+
+
+def get_aws_url():
+    return "s3://" + get_aws_storagepath()
+
+
+def get_gdrive_url():
+    return "gdrive://root/" + str(uuid.uuid4())
+
+
+def get_gcp_storagepath():
+    return TEST_GCP_REPO_BUCKET + "/" + str(uuid.uuid4())
+
+
+def get_gcp_url():
+    return "gs://" + get_gcp_storagepath()
+
+
+def get_azure_url():
+    container_name = os.getenv("AZURE_STORAGE_CONTAINER_NAME")
+    assert container_name is not None
+    return "azure://{}/{}".format(container_name, str(uuid.uuid4()))
+
+
+def get_oss_storagepath():
+    return "{}/{}".format(TEST_OSS_REPO_BUCKET, (uuid.uuid4()))
+
+
+def get_oss_url():
+    return "oss://{}".format(get_oss_storagepath())
+
+
+# NOTE: staticmethod is only needed in Python 2
+class Local:
+    should_test = staticmethod(lambda: True)
+    get_url = staticmethod(get_local_url)
+
+
+class S3:
+    should_test = staticmethod(_should_test_aws)
+    get_url = staticmethod(get_aws_url)
+
+
+class S3Mocked:
+    should_test = staticmethod(lambda: True)
+    get_url = staticmethod(get_aws_url)
+
+    @classmethod
+    @contextmanager
+    def remote(cls):
+        with mock_s3():
+            remote = RemoteS3(None, {"url": cls.get_url()})
+            yield remote
+
+    @classmethod
+    @contextmanager
+    def put_objects(cls, remote, objects):
+        s3 = cls.get_client(remote)
+        s3.create_bucket(Bucket="dvc-test")
+        for key, body in objects.items():
+            cls.put_object(remote, key, body)
+        yield
+
+    @staticmethod
+    def get_client(remote):
+        return remote.s3
+
+    @classmethod
+    def put_object(cls, remote, key, body):
+        s3 = cls.get_client(remote)
+        bucket = remote.path_info.bucket
+
+        s3.put_object(
+            Bucket=bucket, Key=remote.path_info.path + "/" + key, Body=body
+        )
+
+
+class GCP:
+    should_test = staticmethod(_should_test_gcp)
+    get_url = staticmethod(get_gcp_url)
+
+    @classmethod
+    @contextmanager
+    def remote(cls):
+        remote = RemoteGS(None, {"url": cls.get_url()})
+        yield remote
+
+    @classmethod
+    @contextmanager
+    def put_objects(cls, remote, objects):
+        for key, body in objects.items():
+            cls.put_object(remote, key, body)
+        yield
+        cls.remove(remote, objects.keys())
+
+    @classmethod
+    def put_object(cls, remote, key, body):
+        client = cls.get_client(remote)
+        bucket = remote.path_info.bucket
+
+        bucket = client.get_bucket(bucket)
+        blob = bucket.blob(remote.path_info.path + "/" + key)
+        blob.upload_from_string(body)
+
+    @staticmethod
+    def get_client(remote):
+        return remote.gs
+
+    @staticmethod
+    def remove(remote, files):
```

Comment:
We use random prefixes on a bucket, which is cleared periodically, so we shouldn't waste time removing blobs here.

Reply:
ah, didn't know that. Okay, then, there does not need to be any context managers for `put_objects`.